### ISSUE TITLE
Minor error- and some questions

### ISSUE THREAD

**Issue by pythonometrist** [ThilinaRajapakse/pytorch-transformers-classification#1]
Hi there - thanks for putting this together.  Minor issue - you have the train file set to dev.tsv in the data_prep and the colab ipynb.

Second, this might be specific to Windows folk - but using pooling will create issues  if the function is called before  __name__=='__main__' - you might need to add that.

Third, when and if you get a chance - could you shed some light on the number of layers being fine tuned- and if we can specify that - or does this simply add a dense layer with 2 units for binary classification? 

Fourth - XLM seemed to fail out - is there a ready reference for models and vocabs- say bert and  bert_base_cased?  - never mind. Here is a short list https://github.com/huggingface/pytorch-transformers. Also full list https://huggingface.co/pytorch-transformers/pretrained_models.html

Thanks - you are a champ for doing this.

**Comment by ThilinaRajapakse** [comment]
Thanks for pointing out the bugs! I think I fixed the ones you pointed out, plus some others I found along the way.

The `if __name__=__main__:` is a Windows specific issue, which is why I forgot about it. But I added it anyway since it shouldn't affect Linux users either.

I also added a table of reference for model types and model names for convenience.

Please do let me know if you run across any other issues. My GPU has been tied up in a project for a couple of days now, so I haven't been able to test this code as thoroughly as I should have.

The fine-tuning process updates all layers, unlike in most other transfer learning tasks. Which is why you shouldn't use too many training epochs on too small datasets, as these huge models can very easily overfit.

**Comment by pythonometrist** [comment]
First off - thanks for the clarification - I realize that the classification is happening under the hood by the pytorch-transformer package (which is just exposing the head for classification for the corresponding transformer).

There does appear to be a memory leak - I am not an expert in this stuff, but it looks like the model is moved to the gpu multiple times (during training and eval, and again if I use the same approach to report training metrics). , as are the inputs. Do these need to be destroyed using torch.cuda.empty_cache? Essentially as I run cross validation after say 8 iterations I get an OOM issue. Not sure how to destroy the inputs as they seem to be element wise moved to the gpu in each batch.

Interestingly - Within a single run of training and evaluation - OOM occurs on sequence length 50 for more than 32 examples, but accepts 768 examples for inference - I guess the no_grad accounts for this. 

It might help to have a model class - - as of now there are many variables that work fine when using a single script, but run into issues if the functions are split - e.g. args, processors is not a function argument for many functions but is used etc. I am sure you have thought of it though.

Finally - APEX installation on windows is a massive PITA. Any suggestions (though not expected) are most welcome.

Other nitpicks - config is assigned to config_class but never used

**Comment by ThilinaRajapakse** [comment]
Yes, the heavy lifting is being done by the `pytorch-transformer` package. In fact, most of this code is also adapted from the example code provided by Hugging Face.

I can't see how the model will be moved to the GPU multiple times. It's sent to the GPU outside of the training and evaluation functions, and the functions only receive the reference to the model (in the GPU) as a parameter. As for the inputs, you are correct in that they are being sent to the GPU elementwise. The references to those elements are contained as a `tuple` in the `batch` variable. But since both the `batch` and `input` variables are being reassigned at each step, you shouldn't have to delete them manually. They should be garbage collected eventually, although you could call `empty_cache` to make sure that they are immediately scrapped. I'm not sure what the impact would be on training time though, if any.

You can definitely get away with larger batch sizes for inference. But for training, a batch size of 32 is rather big unless you have lots of GPU memory. On a RTX 2080 with `fp16` level 1, I can usually manage a batchsize of 2-8 with sequence lengths 512-1024. Right now, I have a RoBERTa base model running with 512 sequence length, and a batch size of 4 which is using almost all of my 8 GB of memory.

The model itself is in a class imported from `pytorch-transformers`, although I suspect you mean a class for this whole thing;

```
MODEL_CLASSES = {
    'bert': (BertConfig, BertForSequenceClassification, BertTokenizer),
    'xlnet': (XLNetConfig, XLNetForSequenceClassification, XLNetTokenizer),
    'xlm': (XLMConfig, XLMForSequenceClassification, XLMTokenizer),
    'roberta': (RobertaConfig, RobertaForSequenceClassification, RobertaTokenizer)
}
```
It could certainly be done easily, and it might be cleaner to do so. I'll consider making that change. Good point regarding the issue with splitting the functions into separate scripts. I suppose things like `args` and `processors` should at least be optional arguments to the functions that use them.

As for suggestions regarding APEX, I'm afraid the best one I can give is to recommend switching to Linux. It's actually one of the reasons I made the switch permanent. You might have some luck if you built Pytorch from source. But, I never managed to get APEX working on windows.

I left Config there in case it's needed. For example, you need it to convert the model into Torchscript.

**Comment by pythonometrist** [comment]
That helps a lot. 

I guesss I misunderstood the two model.to(device) as two placements of the model to the GPU. My short sequence length is allowing the higher batch size.

Will dig around the OOM more.

The annoying thing is I got 128 gb ram because my 16 core CPU was memory bound - and then I realized GPUs are the way to go and I am back to being memory bound...Scylla and Charibdis of scientific computing.

**Comment by ThilinaRajapakse** [comment]
The second `model.to(device)` call in the evaluation portion is there since we might want to evaluate the model at different checkpoints. So, it loads the model with the weights at each checkpoint and runs the evaluation. But there should only be one model in the GPU at any given time.

You can never have too much RAM! I wish I had 128 gigs. But yeah, GPUs are the priority.

**Comment by ThilinaRajapakse** [comment]
I think the problems with memory are caused by the next model being loaded before the first one can be garbage collected. Maybe it would help to manually delete the model and clear cuda memory before loading the next one. But, manually deleting all references can be tricky since it's rarely enough to delete just the model variable. Other things like results, or training data can hold obscure references to model attributes. Probably the safest way would be to use the notebook for testing and then convert it into a script for the actual training.

**Comment by pythonometrist** [comment]
Agreed - I made them into functions and do the gc before function exits. That helps. Another minor issue - t happended because I am moving between two different gpu set ups - you load_examples functions relies on the position to pick out the labels from the .tsv file. Apparently pandas under different versions will order the dataprep file columns differently - that leads to a label id error. I fixed it manually but I am sure there is a better solution.

**Comment by ThilinaRajapakse** [comment]
Do you mind elaborating on the labelling issue? I don't quite follow what happened. Would be better if you could open a new issue for it, in case other people run into the same problem.

**Comment by pythonometrist** [comment]
Done - I was curious if you can point to an easy starter reference to do my own pre-training  on a custom domain- I tried looking at the pytorch -transformers web page but it looked too overwhelming - shoold I start with the example folder there? Sorrry for the bother!

**Comment by ThilinaRajapakse** [comment]
[This](https://github.com/huggingface/pytorch-transformers/tree/master/examples/lm_finetuning) should be a good point to start. I haven't tried it yet, but I can add something to the repo if and when I do.
